modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -71,39 +72,43 @@ object GlobalSnapshotStorage {
         .compile
         .drain
 
-    Spawn[F].start { offloadProcess }.flatMap { _ =>
-      headRef.get.flatMap {
-        _.traverse { h =>
-          cache(h.ordinal).set(h.some)
-        }
-      }
-    }.map { _ =>
+    def enqueue(snapshot: Signed[GlobalSnapshot]) =
+      cache(snapshot.ordinal).set(snapshot.some) >> snapshot.ordinal
+        .partialPreviousN(inMemoryCapacity)
+        .fold(Applicative[F].unit)(offloadQueue.offer(_))
+
+    Spawn[F].start { offloadProcess }.map { _ =>
       new GlobalSnapshotStorage[F] {
         def prepend(snapshot: Signed[GlobalSnapshot]): F[Boolean] =
-          headRef.modify { maybeCurrent =>
-            maybeCurrent.fold((snapshot.some, true.pure[F])) { current =>
+          headRef.modify {
+            case None =>
+              if (isGenesis(snapshot))
+                (snapshot.some, enqueue(snapshot).map(_ => true))
+              else
+                (none, false.pure[F])
+            case Some(current) =>
               isNextSnapshot(current, snapshot) match {
                 case Left(e) =>
                   (current.some, e.raiseError[F, Boolean])
                 case Right(isNext) if isNext =>
-                  def run =
-                    cache(snapshot.ordinal).set(snapshot.some) >>
-                      snapshot.ordinal
-                        .partialPreviousN(inMemoryCapacity)
-                        .fold(Applicative[F].unit)(offloadQueue.offer)
-
-                  (snapshot.some, run.map(_ => true))
+                  (snapshot.some, enqueue(snapshot).map(_ => true))
 
                 case _ => (current.some, false.pure[F])
               }
-            }
           }.flatten
 
         def head: F[Option[Signed[GlobalSnapshotArtifact]]] = headRef.get
 
         def get(ordinal: SnapshotOrdinal): F[Option[Signed[GlobalSnapshotArtifact]]] = cache(ordinal).get.flatMap {
           case Some(s) => s.some.pure[F]
-          case None    => globalSnapshotLocalFileSystemStorage.read(ordinal).rethrowT.map(_.some)
+          case None =>
+            globalSnapshotLocalFileSystemStorage
+              .read(ordinal)
+              .rethrowT
+              .map(_.some)
+              .recover {
+                case _: NoSuchFileException => none[Signed[GlobalSnapshot]]
```

Comment:
As already mentioned in the online conversation, this should be handled by an abstraction.